Various and sundry

tbowers7 opened the pull request:
A collection of various and sundry minor changes aimed at including more information in product FITS headers and making after-burn scripts easier to call from other code.

Changes:

- Minor bugfix and addition of variable typing in the `pypeit.data.utils` module.
- Add the calibration group information for a frame to the `spec1d` file.  This is placed in the `spec2d` file already, but was not yet propagated to `spec1d`.  This will be important for, _e.g._, automated sensitivity function creation with the new flatfield blazing.
- Fix a typo in an error message in `pypeit.metadata`.
- ~~Add (quiet) options to the `pypeit_sensfunc` script that allow for easier calling from outside code.~~
- Fix the verbosity setting in `pypeit_setup` functionality.
- Add an item to the `SpecObj` datamodel indicating the spatial FWHM of the object in arcsec (the FWHM in pixels is already included).  This value is computed and printed in the `spec1d_*.txt` files, but not included in the `spec1d_*.fits` headers.
- _Add verbosity setting for `pypeit_identify`_

tbowers7 committed:
Update to `SpecObj` datamodel, `SpecObjs` headers, downloading

Add spatial FWHM of extracted objects in arcsec to datamodel:

* The current `specobj` datamodel includes the `FWHM` of the extracted object in
pixels.  In many cases, it would be useful to have this in arcseconds.  This
commit adds an item to the datamodel called `SPAT_FWHM` that is equivalent to
the `opt_fwhm` printed in the `spec1d_*.txt` file.  Use of `OPT_FWHM` is already
taken by the _spectral_ FWHM of lines in the optimal extraction.

* Rethought of naming schemes may be required.

Add calibration association information to spec1d files:

* The calibration associations are included with the spec2d files, but it would
also be useful to include these in the spec1d files for, e.g., pulling the
correct Flat calibration frame for blaze-based sensitivity function computation.

* Since it is not necessary to include this with _each_ SpecObj object within
the SpecObjs list, and SpecObjs does not inherit from `datamodel.DataContainer`,
adding the calibration associations is not as elegant as with the 2D data model.

Add hidden options to `pypeit_sensfunc`:

* Allow for specification of telgrid filename as input to the sensfunc
script, but don't advertise such.  Useful for calling SensFunc from another
python function.

* Allow for specification of polyorder as input to the sensfunc script, but
don't advertise such.  Useful for calling SensFunc from another python function.

Making download more robust:

* Adding more checking and error catching

	modified:   pypeit/data/utils.py
	modified:   pypeit/extraction.py
	modified:   pypeit/pypeit.py
	modified:   pypeit/scripts/sensfunc.py
	modified:   pypeit/specobj.py
	modified:   pypeit/specobjs.py

tbowers7 committed:
Minor cleanup of `pypeit_compare_sky`

	modified:   pypeit/scripts/compare_sky.py

tbowers7 committed:
Typo + verbosity

	modified:   pypeit/metadata.py
	modified:   pypeit/scripts/setup.py

tbowers7 committed:
Merge branch 'develop' into various_and_sundry

kbwestfall commented on pypeit/data/utils.py:
```diff
@@ -494,10 +493,10 @@ def get_extinctfile_filepath(extinction_file: str) -> pathlib.Path:
+    remote_host: str='github',
+    install_script: bool=False,
+    force_update: bool=False,
+    full_url: bool=None,
```
Isn't `full_url` a string?

tbowers7 replied:
Yes, yes it is.

kbwestfall commented on pypeit/metadata.py:
```diff
@@ -893,7 +893,7 @@ def set_configurations(self, configs=None, force=False, fill=None):
                         msgs.warn('When setting the instrument configuration for {0} '.format(ftype)
                                   + 'frames, configuration {0} does not have unique '.format(cfg_key)
-                                  + '{0} values.' .format(meta))
+                                  + '{0} values.' .format(mkey))
```
Agreed that this shouldn't be `meta`, but shouldn't it be `metakey`, not `mkey`?

tbowers7 replied:
I'm not sure.  When I ran this on a data set that triggered this statement, the value of `mkey` displayed made sense.  I'm having a little trouble following the logic in this function, but I chose `mkey` since it is a single value inside a `for` loop, and `metakey` would necessarily be a list.

kbwestfall replied:
Apologies, you're right.  It should be `mkey`.  When looking at the file locally, I was looking at the `release` branch instead of `develop`.

kbwestfall commented on pypeit/scripts/compare_sky.py:
```diff
@@ -25,18 +26,16 @@ def get_parser(cls, width=None):
                             help='Scale user spectrum by a factor')
         parser.add_argument('--test', default=False, action='store_true',
-                            help='Load files but do not show plot')
+                            help=argparse.SUPPRESS)
```
Nice!

kbwestfall commented on pypeit/scripts/sensfunc.py:
```diff
@@ -65,7 +66,8 @@ def get_parser(cls, width=None):
                                  "\nWhere Flat_A_0_DET01.fits is the flat file in your Calibrations directory\n")
-
+        parser.add_argument("--telgridfile", type=str, help=argparse.SUPPRESS)
+        parser.add_argument("--polyorder", type=int, help=argparse.SUPPRESS)
```
I actually wouldn't suppress these.  If you found it helpful, others likely will as well.

tbowers7 replied:
Since these options are more easily settable with the `--sensfile` option when called from the command line, I chose to not muddy the `--help` interface with these.  The purpose of adding them at all is for calling the `SensFunc` class from another python script.

If you still think I should include them in the user-facing interface, I will add proper help strings.

kbwestfall replied:
Is there a reason you're calling this full script from a different python script, instead of calling the `sensfunc.SensFunc` class directly?

tbowers7 replied:
This script sets up the FITS header for the `sensfunc` file and nicely wraps the functionality of the `sensfunc.SensFunc` class.  I was calling the script from a script to automate making many many sensitivity function files for comparison purposes (conditions, stars, airmasses, etc.).  I'm not sure how often users will be doing this kind of thing.  Maybe I just remove these and work through a `--sensfile` like the command-line version.

kbwestfall replied:
Would your script work if you added the same hooks for the telgridfile and polyorder that are in place for, e.g., flatfile?  I.e., essentially make `sensfile` and these other command-line arguments mutually exclusive?  If so, I would expose those command-line arguments, and add the necessary if statements.

tbowers7 replied:
Have removed these options.

kbwestfall commented on pypeit/scripts/setup.py:
```diff
@@ -43,7 +43,7 @@ def get_parser(cls, width=None):
         parser.add_argument('-m', '--manual_extraction', default=False, action='store_true',
                             help='Include the manual extraction column for the user to edit')
-        parser.add_argument('-v', '--verbosity', type=int, default=2,
+        parser.add_argument('-v', '--verbosity', type=int, default=1,
```
Is there a reason to use the level=1 verbosity instead of 2?

tbowers7 replied:
All scripts other than `run_pypeit` have a default verbosity setting of 1.  Also, since this script didn't have the `msgs.set_logfile_and_verbosity()` call yet, it was running with a verbosity of 1 all the time regardless of what the user requested.

kbwestfall replied:
Ah, good point.

kbwestfall reviewed (changes requested):
I have a couple of minor change requests.  Thanks for these edits!

tbowers7 committed:
PR comment

tbowers7 commented:
DevSuite passes.  Note the failures from the first cloud run were caused by a network issue (and the output files were not available for the VET tests):
```
  File "/opt/conda/lib/python3.9/ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out
```

Cloud result:
```
../PypeIt-development-suite/vet_tests/test_flux_coadd1d.py::test_flux_setup_vlt_xshooter FAILED [ 28%]
../PypeIt-development-suite/vet_tests/test_flux_coadd1d.py::test_feige110_ltt3218 FAILED [ 30%]
...
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  234 passed, 73 warnings in 563.02s (0:09:23) ---
--- PYTEST UNIT TESTS PASSED  123 passed, 193 warnings in 1043.15s (0:17:23) ---
--- PYTEST VET TESTS FAILED  2 failed, 51 passed, 34 warnings in 1043.56s (0:17:23) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/215 TESTS  ---
Failed tests:
    vlt_xshooter/VIS_1x1_Feige110 pypeit_sensfunc
Skipped tests:
Testing Started at 2023-08-28T19:22:52.333434
Testing Completed at 2023-08-29T11:59:20.563656
Total Time: 16:36:28.230222
```

After a re-run of the XShooter tests, all DevSuite tests pass, along with the two VET tests from the first run:
```
Reduced data for the following setups:
    vlt_xshooter/VIS_1x1
    vlt_xshooter/VIS_2x1
    vlt_xshooter/VIS_2x2
    vlt_xshooter/VIS_manual
    vlt_xshooter/NIR
    vlt_xshooter/UVB_1x1
    vlt_xshooter/UVB_1x1_Feige110
    vlt_xshooter/VIS_1x1_Feige110
    vlt_xshooter/NIR_Feige110
    vlt_xshooter/VIS_1x1_LTT3218
    vlt_xshooter/NIR_LTT3218

Ran tests in 8 parallel processes
...
../PypeIt-development-suite/vet_tests/test_flux_coadd1d.py::test_flux_setup_vlt_xshooter PASSED [ 28%]
../PypeIt-development-suite/vet_tests/test_flux_coadd1d.py::test_feige110_ltt3218 PASSED [ 30%]
...
--------------------------------------------------------
--- PYTEST VET TESTS FAILED  47 failed, 6 passed, 36 warnings in 245.93s (0:04:05) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 17/17 TESTS  ---
Testing Started at 2023-08-29T20:30:46.178474
Testing Completed at 2023-08-29T21:29:17.081173
Total Time: 0:58:30.902699

```

tbowers7 committed:
Add verbosity option to `pypeit_identify`

It wants to be like all the other cool cats.

	modified:   pypeit/scripts/identify.py

tbowers7 committed:
Revert changes in `pypeit_sensfunc`

	modified:   pypeit/scripts/sensfunc.py

kbwestfall reviewed (approved):
Thanks, Tim!

debora-pe commented on pypeit/specobjs.py:
```diff
@@ -752,6 +753,10 @@ def write_to_fits(self, subheader, outfile, overwrite=True, update_det=None,
+        # Add calibration associations to Header
+        if self.calibs is not None:
+            for key, val in self.calibs.items():
+                header[f'CLBS_{key}'] = val
```
I think it would be good to add also the possibility to load the calibration association from the header, as done for spec2d here: https://github.com/pypeit/PypeIt/blob/9931a1e356ce9c0ca2dc105ebeeab63419b5e441/pypeit/spec2dobj.py#L177-L183 
I'm not sure where it would go here, maybe in `from_fitsfile()`.
What do you think @kbwestfall ?

kbwestfall replied:
Yes, good idea!

tbowers7 replied:
Added.

The new change passes the repo unit tests, but I haven't rerun the DevSuite.  How do you want to proceed?

debora-pe replied:
@tbowers7 Thank you for doing that. I think it would be safe to re-run the Dev Suite. If you have problems, I can run it this evening on my machine.

tbowers7 replied:
@debora-pe , I no longer have access to the cloud and don't have a machine locally with enough brawn to run the DevSuite.  So, if you could have a go at it, I would appreciate it.

debora-pe replied:
@tbowers7 OK, I'll do that. I'll post the results here.

debora-pe reviewed (changes requested):
@tbowers7 all looks good. I only have a comment about loading the calibration association from a spec1d file.

tbowers7 committed:
Read in calibration associations from SpecObjs FITS files

	modified:   pypeit/specobjs.py

tbowers7 committed:
Merge remote-tracking branch 'upstream/develop' into various_and_sundry

tbowers7 committed:
Align `verbosity` help with other scripts.

	modified:   pypeit/scripts/setup.py

debora-pe commented:
@tbowers7 All good with the tests run. There is a failure in the vet test, but I think it's related to another PR.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  234 passed, 72 warnings in 312.84s (0:05:12) ---
--- PYTEST UNIT TESTS PASSED  123 passed, 195 warnings in 1536.97s (0:25:36) ---
--- PYTEST VET TESTS FAILED  1 failed, 52 passed, 69 warnings in 569.08s (0:09:29) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 217/217 TESTS  ---
Testing Started at 2023-09-05T20:58:31.154145
Testing Completed at 2023-09-06T20:05:58.378193
Total Time: 23:07:27.224048
```

debora-pe reviewed (approved):
Thanks!

profxj commented:
Feel free to merge
